# `pandas` Part 7: Combining Datasets with `concat()`

# Learning Objectives
## By the end of this tutorial you will be able to:
1. Combine DataFrames and/or Series with `concat()`
2. Understand a multi-index
3. Reset an index with `reset_index()`
4. Perform descriptive analytics on a combined DataFrame

## Files Needed for this lesson:
>- `CAvideos.csv`
>- `GBvideos.csv`
>- Download this csv files from Canvas prior to the lesson
>- C:\\Users\\mimc2537\\OneDrive - UCB-O365\\python\\pandas\\

## The general steps to working with pandas:
1. import pandas as pd
2. Create or load data into a pandas DataFrame or Series
3. Reading data with `pd.read_`
>- Excel files: `pd.read_excel('fileName.xlsx')`
>- Csv files: `pd.read_csv('fileName.csv')`
>- Note: if the file you want to read into your notebook is not in the same folder you can do one of two things:
>>- Move the file you want to read into the same folder/directory as the notebook
>>- Type out the full path into the read function
4. After steps 1-3 you will want to check out your DataFrame
>- Use `shape` to see how many records and columns are in your DataFrame
>- Use `head()` to show the first 5-10 records in your DataFrame

# Introduction Notes on Combining Data Using `pandas`
1. Being able to combine data from multiple sources is a critical skill for analytics professionals
2. We will learn the `pandas` way of combining data but there are similarities here to SQL
3. Why combine data with `pandas` if you can do the same thing in SQL?
>- The answer to this depends on the project
>- Some projects may be completed more efficiently all with `pandas` so you wouldn't necessarily need SQL
>- For some projects incorporating SQL into our python code makes sense
>- In a an analytics job, you will likely use both python and SQL to get the job done! 

# Initial set-up steps
1. import modules and check working directory
2. Read data in
3. Check the data

In [1]:
import pandas as pd, os
os.getcwd()

'C:\\Users\\Cupcake\\Google Drive\\School\\Senior Year 2020-2021\\Fall 2020\\BAIM 3220\\Python\\Week12'

# Step 2 Read Data Into a DataFrame with `read_csv()`
>- file names: 
>>- `CAvideos.csv`
>>- `GBvideos.csv`

In [3]:
can_youtube = pd.read_csv('CAvideos.csv')
uk_youtube = pd.read_csv('GBvideos.csv')

### Check how many rows and columns are in our DataFrames

In [6]:
print(can_youtube.shape,uk_youtube.shape)

(40881, 16) (38916, 16)


### Check a couple of rows of data in one of the new DataFrames

In [7]:
can_youtube.head()

video_id trending_date  \
0  n1WpP7iowLc      17.14.11   
1  0dBIkQ4Mz1M      17.14.11   
2  5qpjK5DgCt4      17.14.11   
3  d380meD0W0M      17.14.11   
4  2Vv-BfVoq4g      17.14.11   

                                               title channel_title  \
0         Eminem - Walk On Water (Audio) ft. Beyoncé    EminemVEVO   
1                      PLUSH - Bad Unboxing Fan Mail     iDubbbzTV   
2  Racist Superman | Rudy Mancuso, King Bach & Le...  Rudy Mancuso   
3                           I Dare You: GOING BALD!?      nigahiga   
4        Ed Sheeran - Perfect (Official Music Video)    Ed Sheeran   

   category_id              publish_time  \
0           10  2017-11-10T17:00:03.000Z   
1           23  2017-11-13T17:00:00.000Z   
2           23  2017-11-12T19:05:24.000Z   
3           24  2017-11-12T18:01:41.000Z   
4           10  2017-11-09T11:04:14.000Z   

                                                tags     views    likes  \
0  Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579   787425   
1  plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   1014651   127794   
2  racist superman|"rudy"|"mancuso"|"king"|"bach"...   3191434   146035   
3  ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...   2095828   132239   
4  edsheeran|"ed sheeran"|"acoustic"|"live"|"cove...  33523622  1634130   

   dislikes  comment_count                                  thumbnail_link  \
0     43420         125882  https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
1      1688          13030  https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg   
2      5339           8181  https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg   
3      1989          17518  https://i.ytimg.com/vi/d380meD0W0M/default.jpg   
4     21082          85067  https://i.ytimg.com/vi/2Vv-BfVoq4g/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   
2              False             False                   False   
3              False             False                   False   
4              False             False                   False   

                                         description  
0  Eminem's new track Walk on Water ft. Beyoncé i...  
1  STill got a lot of packages. Probably will las...  
2  WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...  
3  I know it's been a while since we did this sho...  
4  🎧: https://ad.gt/yt-perfect\n💰: https://atlant...

## Check the datatypes

In [8]:
can_youtube.dtypes

video_id                  object
trending_date             object
title                     object
channel_title             object
category_id                int64
publish_time              object
tags                      object
views                      int64
likes                      int64
dislikes                   int64
comment_count              int64
thumbnail_link            object
comments_disabled           bool
ratings_disabled            bool
video_error_or_removed      bool
description               object
dtype: object

# Combining DataFrames
>- The three common ways to combine datastest in pandas is with `concat()`, `join()`, and `merge()`
>- `concat()` will take two DataFrames or Series and append them together
>>- This is basically taking DataFrames and stacking their data on top of each other into one DataFrame
>>- For `concat()` you need the columns/fields in both DataFrames to the be the same
>- `join()` "links" DataFrames together based on a common field/column between the two
>- `merge()` also links DataFrames together based on common field/columns but with different syntax.
>>- We will cover the most basic join in this class
>>- A more in depth study of joins is provided in SQL focused courses
>>- Pandas join reference for further study: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.join.html


# Using the YouTube DataFrames to practice combining data with pandas
>- The YouTube datasets store data on various YouTube trending statistics
>- Our example datasets show several months of data and daily trending YouTube videos.

>- For more information and for other YouTube datasets see the following link:
>>- https://www.kaggle.com/datasnaek/youtube-new

### First, creating a new DataFrame that appends the Canadian and British YouTube DataFrames

In [11]:
canUk = pd.concat([can_youtube,uk_youtube],
                                  keys=['can','uk'],
                                  names = ['country','rowid_country'])
canUk

video_id trending_date  \
country rowid_country                              
can     0              n1WpP7iowLc      17.14.11   
        1              0dBIkQ4Mz1M      17.14.11   
        2              5qpjK5DgCt4      17.14.11   
        3              d380meD0W0M      17.14.11   
        4              2Vv-BfVoq4g      17.14.11   
...                            ...           ...   
uk      38911          l884wKofd54      18.14.06   
        38912          IP8k2xkhOdI      18.14.06   
        38913          Il-an3K9pjg      18.14.06   
        38914          -DRsfNObKIQ      18.14.06   
        38915          4YFo4bdMO8Q      18.14.06   

                                                                   title  \
country rowid_country                                                      
can     0                     Eminem - Walk On Water (Audio) ft. Beyoncé   
        1                                  PLUSH - Bad Unboxing Fan Mail   
        2              Racist Superman | Rudy Mancuso, King Bach & Le...   
        3                                       I Dare You: GOING BALD!?   
        4                    Ed Sheeran - Perfect (Official Music Video)   
...                                                                  ...   
uk      38911          Enrique Iglesias - MOVE TO MIAMI (Official Vid...   
        38912          Jacob Sartorius - Up With It (Official Music V...   
        38913                         Anne-Marie - 2002 [Official Video]   
        38914          Eleni Foureira - Fuego - Cyprus - LIVE - First...   
        38915          KYLE - Ikuyo feat.  2 Chainz & Sophia Black [A...   

                                 channel_title  category_id  \
country rowid_country                                         
can     0                           EminemVEVO           10   
        1                            iDubbbzTV           23   
        2                         Rudy Mancuso           23   
        3                             nigahiga           24   
        4                           Ed Sheeran           10   
...                                        ...          ...   
uk      38911              EnriqueIglesiasVEVO           10   
        38912                  Jacob Sartorius           10   
        38913                       Anne-Marie           10   
        38914          Eurovision Song Contest           24   
        38915                   SuperDuperKyle           10   

                                   publish_time  \
country rowid_country                             
can     0              2017-11-10T17:00:03.000Z   
        1              2017-11-13T17:00:00.000Z   
        2              2017-11-12T19:05:24.000Z   
        3              2017-11-12T18:01:41.000Z   
        4              2017-11-09T11:04:14.000Z   
...                                         ...   
uk      38911          2018-05-09T07:00:01.000Z   
        38912          2018-05-11T17:09:16.000Z   
        38913          2018-05-08T11:05:08.000Z   
        38914          2018-05-08T20:32:32.000Z   
        38915          2018-05-11T04:06:35.000Z   

                                                                    tags  \
country rowid_country                                                      
can     0              Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...   
        1              plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   
        2              racist superman|"rudy"|"mancuso"|"king"|"bach"...   
        3              ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...   
        4              edsheeran|"ed sheeran"|"acoustic"|"live"|"cove...   
...                                                                  ...   
uk      38911          Enrique Iglesias feat. Pitbull|"MOVE TO MIAMI"...   
        38912          jacob sartorius|"jacob"|"up with it"|"jacob sa...   
        38913          anne|"marie"|"anne-marie"|"2002"|"two thousand...   
        38914          Eurovision Song C

#### Some notes on the previous code
>- Line 1: We define a new DataFrame named `CanUK` which is defined as the concatenation,`concat()`, of two datasets
>>- Dataset 1 = canadian_youtube
>>- Dataset 2 = uk_youtube
>>- The `concat()` function takes the two (or more if applicable) DataFrames and "stacks" them on top of each
>- Line 2: We use `keys` option to define a multi-index (aka hierarchical index)
>>- Because our datasets represent YouTube videos from different countries we pass the abbreviated names of those countries as a list to `keys`
>>- Enter the keys names in order they appear in line 1 (e.g., 'can' first, 'uk' second)
>- Line 3: We use the `names` option to label our index columns from line 2
>>- Without the `names` option we would not have anything above our index columns

### Check the index for any dataframe using `DataFrame.index`
>- Note how `concat()` uses the rowid's for each country's dataset versus continuing the count

In [13]:
canUk.index

MultiIndex([('can',     0),
            ('can',     1),
            ('can',     2),
            ('can',     3),
            ('can',     4),
            ('can',     5),
            ('can',     6),
            ('can',     7),
            ('can',     8),
            ('can',     9),
            ...
            ( 'uk', 38906),
            ( 'uk', 38907),
            ( 'uk', 38908),
            ( 'uk', 38909),
            ( 'uk', 38910),
            ( 'uk', 38911),
            ( 'uk', 38912),
            ( 'uk', 38913),
            ( 'uk', 38914),
            ( 'uk', 38915)],
           names=['country', 'rowid_country'], length=79797)

### Take a look at our new DataFrame

In [14]:
canUk.head()

video_id trending_date  \
country rowid_country                              
can     0              n1WpP7iowLc      17.14.11   
        1              0dBIkQ4Mz1M      17.14.11   
        2              5qpjK5DgCt4      17.14.11   
        3              d380meD0W0M      17.14.11   
        4              2Vv-BfVoq4g      17.14.11   

                                                                   title  \
country rowid_country                                                      
can     0                     Eminem - Walk On Water (Audio) ft. Beyoncé   
        1                                  PLUSH - Bad Unboxing Fan Mail   
        2              Racist Superman | Rudy Mancuso, King Bach & Le...   
        3                                       I Dare You: GOING BALD!?   
        4                    Ed Sheeran - Perfect (Official Music Video)   

                      channel_title  category_id              publish_time  \
country rowid_country                                                        
can     0                EminemVEVO           10  2017-11-10T17:00:03.000Z   
        1                 iDubbbzTV           23  2017-11-13T17:00:00.000Z   
        2              Rudy Mancuso           23  2017-11-12T19:05:24.000Z   
        3                  nigahiga           24  2017-11-12T18:01:41.000Z   
        4                Ed Sheeran           10  2017-11-09T11:04:14.000Z   

                                                                    tags  \
country rowid_country                                                      
can     0              Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...   
        1              plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   
        2              racist superman|"rudy"|"mancuso"|"king"|"bach"...   
        3              ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...   
        4              edsheeran|"ed sheeran"|"acoustic"|"live"|"cove...   

                          views    likes  dislikes  comment_count  \
country rowid_country                                               
can     0              17158579   787425     43420         125882   
        1               1014651   127794      1688          13030   
        2               3191434   146035      5339           8181   
        3               2095828   132239      1989          17518   
        4              33523622  1634130     21082          85067   

                                                       thumbnail_link  \
country rowid_country                                                   
can     0              https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
        1              https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg   
        2              https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg   
        3              https://i.ytimg.com/vi/d380meD0W0M/default.jpg   
        4              https://i.ytimg.com/vi/2Vv-BfVoq4g/default.jpg   

                       comments_disabled  ratings_disabled  \
country rowid_country                                        
can     0                          False             False   
        1                          False             False   
        2                          False             False   
        3                          False             False   
        4                          False             False   

                       video_error_or_removed  \
country rowid_country                           
can     0                               False   
        1                               False   
        2                               False   
        3                               False   
        4                               False   

                                                             description  
country rowid_country                                                     
can     0              Eminem's new track Walk on Water ft. Beyoncé i...  
        1              STill got a lot of packag

In [18]:
canUk.tail()

video_id trending_date  \
country rowid_country                              
uk      38911          l884wKofd54      18.14.06   
        38912          IP8k2xkhOdI      18.14.06   
        38913          Il-an3K9pjg      18.14.06   
        38914          -DRsfNObKIQ      18.14.06   
        38915          4YFo4bdMO8Q      18.14.06   

                                                                   title  \
country rowid_country                                                      
uk      38911          Enrique Iglesias - MOVE TO MIAMI (Official Vid...   
        38912          Jacob Sartorius - Up With It (Official Music V...   
        38913                         Anne-Marie - 2002 [Official Video]   
        38914          Eleni Foureira - Fuego - Cyprus - LIVE - First...   
        38915          KYLE - Ikuyo feat.  2 Chainz & Sophia Black [A...   

                                 channel_title  category_id  \
country rowid_country                                         
uk      38911              EnriqueIglesiasVEVO           10   
        38912                  Jacob Sartorius           10   
        38913                       Anne-Marie           10   
        38914          Eurovision Song Contest           24   
        38915                   SuperDuperKyle           10   

                                   publish_time  \
country rowid_country                             
uk      38911          2018-05-09T07:00:01.000Z   
        38912          2018-05-11T17:09:16.000Z   
        38913          2018-05-08T11:05:08.000Z   
        38914          2018-05-08T20:32:32.000Z   
        38915          2018-05-11T04:06:35.000Z   

                                                                    tags  \
country rowid_country                                                      
uk      38911          Enrique Iglesias feat. Pitbull|"MOVE TO MIAMI"...   
        38912          jacob sartorius|"jacob"|"up with it"|"jacob sa...   
        38913          anne|"marie"|"anne-marie"|"2002"|"two thousand...   
        38914          Eurovision Song Contest|"2018"|"Lisbon"|"Cypru...   
        38915          Kyle|"SuperDuperKyle"|"Ikuyo"|"2 Chainz"|"Soph...   

                          views   likes  dislikes  comment_count  \
country rowid_country                                              
uk      38911          25066952  268088     12783           9933   
        38912           1492219   61998     13781          24330   
        38913          29641412  394830      8892          19988   
        38914          14317515  151870     45875          26766   
        38915            607552   18271       274           1423   

                                                       thumbnail_link  \
country rowid_country                                                   
uk      38911          https://i.ytimg.com/vi/l884wKofd54/default.jpg   
        38912          https://i.ytimg.com/vi/IP8k2xkhOdI/default.jpg   
        38913          https://i.ytimg.com/vi/Il-an3K9pjg/default.jpg   
        38914          https://i.ytimg.com/vi/-DRsfNObKIQ/default.jpg   
        38915          https://i.ytimg.com/vi/4YFo4bdMO8Q/default.jpg   

                       comments_disabled  ratings_disabled  \
country rowid_country                                        
uk      38911                      False             False   
        38912                      False             False   
        38913                      False             False   
        38914                      False             False   
        38915                      False             False   

                       video_error_or_removed  \
country rowid_country                           
uk      38911                           False   
        38912                           False   
        38913                           False   
        38914                           False   
        38915                           False   

                           

In [19]:
canUk.shape

(79797, 16)

#### Did using `concat()` work to append the two DataFrames together? 
>- Check the shape of your new DataFrame
>- Compare the number of records to each one individually
>>- canadian_youtube = 40881 records
>>- uk_youtube = 38916 records
>>- 40881 + 38916 = 79797 total records

#### `reset_index`:
##### Note: You can reset a an index with `reset_index` 
>- This can be useful for some situations
>- For a multi-index you can pass the `level` option and specify what index you want to reset
>- Note: To make the change to our current DataFrame we would need to use the option, `inplace=True`

In [22]:
canUk.reset_index(level='country').head(2)
# canUk.reset_index(level='country',inplace=True).head(2)

country     video_id trending_date  \
rowid_country                                      
0                 can  n1WpP7iowLc      17.14.11   
1                 can  0dBIkQ4Mz1M      17.14.11   

                                                    title channel_title  \
rowid_country                                                             
0              Eminem - Walk On Water (Audio) ft. Beyoncé    EminemVEVO   
1                           PLUSH - Bad Unboxing Fan Mail     iDubbbzTV   

               category_id              publish_time  \
rowid_country                                          
0                       10  2017-11-10T17:00:03.000Z   
1                       23  2017-11-13T17:00:00.000Z   

                                                            tags     views  \
rowid_country                                                                
0              Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579   
1              plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   1014651   

                likes  dislikes  comment_count  \
rowid_country                                    
0              787425     43420         125882   
1              127794      1688          13030   

                                               thumbnail_link  \
rowid_country                                                   
0              https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
1              https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg   

               comments_disabled  ratings_disabled  video_error_or_removed  \
rowid_country                                                                
0                          False             False                   False   
1                          False             False                   False   

                                                     description  
rowid_country                                                     
0              Eminem's new track Walk on Water ft. Beyoncé i...  
1              STill got a lot of packages. Probably will las...

In [23]:
canUk.head(2)

video_id trending_date  \
country rowid_country                              
can     0              n1WpP7iowLc      17.14.11   
        1              0dBIkQ4Mz1M      17.14.11   

                                                            title  \
country rowid_country                                               
can     0              Eminem - Walk On Water (Audio) ft. Beyoncé   
        1                           PLUSH - Bad Unboxing Fan Mail   

                      channel_title  category_id              publish_time  \
country rowid_country                                                        
can     0                EminemVEVO           10  2017-11-10T17:00:03.000Z   
        1                 iDubbbzTV           23  2017-11-13T17:00:00.000Z   

                                                                    tags  \
country rowid_country                                                      
can     0              Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...   
        1              plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   

                          views   likes  dislikes  comment_count  \
country rowid_country                                              
can     0              17158579  787425     43420         125882   
        1               1014651  127794      1688          13030   

                                                       thumbnail_link  \
country rowid_country                                                   
can     0              https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
        1              https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg   

                       comments_disabled  ratings_disabled  \
country rowid_country                                        
can     0                          False             False   
        1                          False             False   

                       video_error_or_removed  \
country rowid_country                           
can     0                               False   
        1                               False   

                                                             description  
country rowid_country                                                     
can     0              Eminem's new track Walk on Water ft. Beyoncé i...  
        1              STill got a lot of packages. Probably will las...

# Now some descriptive analytics

### What channels have the most trending videos?

In [24]:
canUk.channel_title.value_counts()

The Late Show with Stephen Colbert        361
TheEllenShow                              357
Jimmy Kimmel Live                         351
The Tonight Show Starring Jimmy Fallon    345
Late Night with Seth Meyers               342
                                         ... 
King Z                                      1
Cartoon Network                             1
Nash Grier                                  1
Pero Like                                   1
Gong Bao                                    1
Name: channel_title, Length: 5956, dtype: int64

### What are the quantitative descriptive statistics for TheEllenShow?

In [30]:
canUk.where(canUk.channel_title == 'TheEllenShow').describe()

category_id         views          likes      dislikes  comment_count  \
count        357.0  3.570000e+02     357.000000    357.000000     357.000000   
mean          24.0  1.779820e+06   46257.770308   1328.537815    2179.221289   
std            0.0  2.008212e+06   62550.222512   3198.710961    3614.375611   
min           24.0  5.631300e+04    1187.000000     18.000000       0.000000   
25%           24.0  5.360520e+05   14152.000000    243.000000     318.000000   
50%           24.0  1.238747e+06   28108.000000    598.000000     940.000000   
75%           24.0  2.022094e+06   59950.000000   1309.000000    2858.000000   
max           24.0  1.359265e+07  499673.000000  26397.000000   28381.000000   

       comments_disabled  ratings_disabled  video_error_or_removed  
count         357.000000             357.0                   357.0  
mean            0.170868               0.0                     0.0  
std             0.376922               0.0                     0.0  
min             0.000000               0.0                     0.0  
25%             0.000000               0.0                     0.0  
50%             0.000000               0.0                     0.0  
75%             0.000000               0.0                     0.0  
max             1.000000               0.0                     0.0

##### Alternatively, you can use `loc[]` to peform the filtering operation
>- The use of `where()` or `loc[]` depends on the question/purpose or sometimes just personal preference

In [32]:
canUk.loc[canUk['channel_title'] == 'TheEllenShow'].describe()

category_id         views          likes      dislikes  comment_count
count        357.0  3.570000e+02     357.000000    357.000000     357.000000
mean          24.0  1.779820e+06   46257.770308   1328.537815    2179.221289
std            0.0  2.008212e+06   62550.222512   3198.710961    3614.375611
min           24.0  5.631300e+04    1187.000000     18.000000       0.000000
25%           24.0  5.360520e+05   14152.000000    243.000000     318.000000
50%           24.0  1.238747e+06   28108.000000    598.000000     940.000000
75%           24.0  2.022094e+06   59950.000000   1309.000000    2858.000000
max           24.0  1.359265e+07  499673.000000  26397.000000   28381.000000

### What were the total YouTube videos, total views, likes and dislikes for TheEllenShow?
>- Using the agg() function to calculate specific aggregations on different columns

In [38]:
canUk.where(canUk['channel_title'] == 'TheEllenShow').agg({'title':['count'],
                                                          'views':['sum'],
                                                          'likes':['sum'],
                                                          'dislikes':['sum']})

title        views       likes  dislikes
count  357.0          NaN         NaN       NaN
sum      NaN  635395636.0  16514024.0  474288.0

In [37]:
canUk.loc[canUk['channel_title'] == 'TheEllenShow'].agg({'title':['count'],
                                                          'views':['sum'],
                                                          'likes':['sum'],
                                                          'dislikes':['sum']})

title        views       likes  dislikes
count  357.0          NaN         NaN       NaN
sum      NaN  635395636.0  16514024.0  474288.0

## What are the totals for TheEllenShow's top 5 most viewed videos?
>- Only include the title names as part of the output (not channel or any other categorical fields)
>- Include total views, likes, dislikes, and comment count in the output

In [41]:
canUk.loc[canUk['channel_title'] == 'TheEllenShow',
             ['title','views','likes','dislikes','comment_count']].\
        groupby(['title']).sum().\
        sort_values(by='views',ascending=False).\
        head()

views    likes  \
title                                                                   
Jennifer Lawrence Explains Her Drunk Alter Ego ...  71621685  1302161   
Kid Yodeler Mason Ramsey Performs                   68553312  2697959   
Kim Kardashian Lets Gender of Third Child Slip      52985123   776177   
Billionaire Bill Gates Guesses Grocery Store Pr...  37076353   542804   
Laurel or Yanny?                                    28695358   613487   

                                                    dislikes  comment_count  
title                                                                        
Jennifer Lawrence Explains Her Drunk Alter Ego ...     35255          49762  
Kid Yodeler Mason Ramsey Performs                     144604              0  
Kim Kardashian Lets Gender of Third Child Slip         26509          50940  
Billionaire Bill Gates Guesses Grocery Store Pr...     29036          42720  
Laurel or Yanny?                                       18568         100731

In [58]:
canUk[['title','views','likes','dislikes','comment_count']].where(canUk.channel_title == 'TheEllenShow').\
        groupby(['title']).sum().sort_values(by='views',ascending=False).head()

views      likes  \
title                                                                       
Jennifer Lawrence Explains Her Drunk Alter Ego ...  71621685.0  1302161.0   
Kid Yodeler Mason Ramsey Performs                   68553312.0  2697959.0   
Kim Kardashian Lets Gender of Third Child Slip      52985123.0   776177.0   
Billionaire Bill Gates Guesses Grocery Store Pr...  37076353.0   542804.0   
Laurel or Yanny?                                    28695358.0   613487.0   

                                                    dislikes  comment_count  
title                                                                        
Jennifer Lawrence Explains Her Drunk Alter Ego ...   35255.0        49762.0  
Kid Yodeler Mason Ramsey Performs                   144604.0            0.0  
Kim Kardashian Lets Gender of Third Child Slip       26509.0        50940.0  
Billionaire Bill Gates Guesses Grocery Store Pr...   29036.0        42720.0  
Laurel or Yanny?                                     18568.0       100731.0

# Some Notes on the Previous Example
>- Our pandas code in the previous example is similar to SQL in the following ways
    1. `loc[CanUk.channel_title == 'TheEllenShow',` is SQL equivalent to `WHERE channel_title = 'TheEllenShow'`
    2. `['title','views','likes','dislikes','comment_count']` is SQL equivalent to:
        `SELECT title, sum(views),sum(likes),sum(dislikes),sum(comment_count)`
    3. `groupby(['title`]) is SQL equivalent to GROUP BY title
    4. Now in pandas we enter the aggregation after the `groupby()`, in this example `sum()`
      >>- In SQL we write the aggregation in the SELECT statement
      
## In future lessons we will continue to learn how pandas and SQL relate

In [14]:
import mechanize

url = input("Enter the full url")

request = mechanize.Browser()
request.open(url)

request.select_form(nr = 0)

request["id"] = "1" #or 1 = "1"

response = request.submit()
content = response.read()
print(content)

import mechanize

url = input("Enter the full url")
attack_no = 1

with open ('vectors.txt') as v:

    for line in v:
       browser.open(url)
       browser.select_form(nr = 0)
       browser['id'] = line
       res = browser.submit()
    content = res.read()

output = open('response/' + str(attack_no) + '.txt', 'w')
output.write(content)
output.close()
print(attack_no)
attack_no += 1

ModuleNotFoundError: No module named 'mechanize'